src/gradients.rs

```diff
@@ -52,7 +54,15 @@ impl<E: Unit, D: DeviceStorage> Gradients<E, D> {
 
     /// Drops all gradients except for the ids specified in the parameter.
     pub fn retain(&mut self, ids: &[UniqueId]) {
-        self.gradient_by_id.retain(|k, _| ids.contains(k));
+        self.non_temporary.extend(ids);
+        self.drop_temporaries();
+    }
+
+    /// Keeps all gradients marked previously by [Gradients::retain], and drops all
+    /// others.
+    pub fn drop_temporaries(&mut self) {
+        self.gradient_by_id
+            .retain(|k, _| self.non_temporary.contains(k));
```

Comment:
Needs to handle when retain hasn't been called (i.e. all gradients are temporary)